### PR TITLE
fix endless loop trying to connect to a specific shard

### PR DESCRIPTION
**Pull request by thirstycrow**
Endless loop occurred when connecting to one of our scylla cluster, which is shared among serveral application, and some of which is still using the cassandra driver, so the connection numbers on each shard varies. 

When scylla driver trying to create a connection to some shard A that has at least 2 more connections than some other shard B in the following loop, it can never success. Because on the server side the new connection is always transferred to shard B, and 
on the client side the new connection, which is connected to shard B, will be closed immediately, if there is already a connection to the same shard in the `connectionsToClose` map, so the connection count on shard B is always less than that on shard A, thus endless loop. In consequence, lots of closed connections stays in TIME_WAIT state, and new connections cannot be established due to exhaust of socket resources.

https://github.com/scylladb/java-driver/blob/3aec70da8bc927afdf69ca455dfd05f81a8bb53e/driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java#L670-L677

To fix the problem the `connectionsToClose` is changed to be able to hold more than one connection for each shard. As new connections created, connection count on shard B can eventually be equal to or even more than that on shard A.

### CODE REVIEW

**Comment by haaawk** [comment]
Hello @thirstycrow. Thank you for your contribution but the solution is not going to be that easy I'm afraid. The driver was already operating in this unbounded manner before and we had to limit the number of connections hold. With global clusters when multiple clients were connecting to the same cluster, each client keeping an unlimited number of unused connections led to the exhaustion of socket resources.

We've recently added a capability of routing connections to a selected shard in Scylla: https://github.com/scylladb/scylla/pull/6961 and we will be migrating Java driver to it soon. This will solve all the problems related to not being able to hit the right shard.

**Comment by thirstycrow** [comment]
@haaawk Thank you for reply. It's good to know this nice solution. The migration you mentioned is to make the Java driver auto discover the scylla shard aware port, but the released version should work well as long as configured to use the new port, is it right?

**Comment by haaawk** [comment]
There are two dimensions here:
1. Java driver can support the new approach or not
2. Scylla server can support the new approach or not (only Scylla 4.3 and higher will have the support)

There are 4 combinations of those possibilities and our intent is to keep all of them working but only the case when both Java driver and Scylla server support the new approach, the new approach will work and show its benefits.

**Comment by thirstycrow** [comment]
We tested the shard ware port feature with the released scylla driver, the results showed that connecting rate was very high from the beginning and dropped slow to zero at 20 minutes after the server started. We made the following changes to the driver, and now it worked smoothly with both shard aware and non shard aware ports. 
1. Background connecting rate is limited to 1 per second per `HostConnectionPool`.
2. A client port is picked according to the target shard id and shards count before connecting to an advertised shard aware port.
3. Always connect to the shard aware port, even if it differs from the configured port used on initialization.

**Comment by haaawk** [comment]
Thanks, @thirstycrow. I will try to review the PR soon.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 1]
```diff
@@ -86,10 +89,13 @@
 
   private AtomicInteger[] scheduledForCreation;
 
-  private final EventExecutor timeoutsExecutor;
+  private final EventExecutor scheduler;
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:93]: Please don't rename variables like that. We're rebasing the driver on top of Cassandra driver every now and then. Renames like this can cause conflicts and make it difficult to sync with Cassandra driver.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 2]
```diff
@@ -86,10 +89,13 @@
...
 
   private final AtomicReference<CloseFuture> closeFuture = new AtomicReference<CloseFuture>();
 
+  private final BlockingQueue<ConnectionTask> pendingTasks =
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:97]: Please don't mix two unrelated changes in one commit. Rate limiting should be one commit and using local port should be another.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 3]
```diff
@@ -86,10 +89,13 @@
...
+      new LinkedBlockingQueue<ConnectionTask>();
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:98]: no need to repeat collection type `<>` is enough.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 4]
```diff
@@ -171,15 +183,33 @@ public void run() {
   // following threshold, we just replace the connection by a new one.
   private final int minAllowedStreams;
 
-  HostConnectionPool(Host host, HostDistance hostDistance, SessionManager manager) {
+  HostConnectionPool(Host host, HostDistance hostDistance, final SessionManager manager) {
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:187]: Why is this final needed?

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 5]
```diff
@@ -171,15 +183,33 @@ public void run() {
...
+
+    scheduler.schedule(
+        new Runnable() {
+          @Override
+          public void run() {
+            if (!isClosed()) {
+              ConnectionTask task = pendingTasks.poll();
+              if (task != null) {
+                manager.blockingExecutor().submit(task);
+                scheduler.schedule(this, 1, TimeUnit.SECONDS);
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:205]: Instead of having an infinite loop like that we can schedule tasks directly on the scheduler with 100ms delay.

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 6]
```diff
@@ -406,26 +436,26 @@ private Connection findLeastBusyForShard(int shardId) {
     }
 
     Connection leastBusy = null;
+    int lastCheckedShard = shardId;
 
     if (connections[shardId].isEmpty()) {
       if (host.convictionPolicy.canReconnectNow()) {
         if (connectionsPerShard == 0) {
           maybeSpawnNewConnection(shardId);
-          return enqueue(timeout, unit, maxQueueSize, shardId);
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:414]: Why is this and line 421 removed?

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 7]
```diff
@@ -451,7 +481,7 @@ private Connection findLeastBusyForShard(int shardId) {
             >= Math.min(
                 leastBusy.maxAvailableStreams(),
                 options().getMaxRequestsPerConnection(hostDistance))) {
-          return enqueue(timeout, unit, maxQueueSize, shardId);
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:454]: I think this is wrong. If we can't find connection we should enqueue for the requested shardId

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java (hunk 6 above)
Comment (haaawk) [reply driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:414]: Ok. I see it now

File: driver-core/src/main/java/com/datastax/driver/core/ShardingInfo.java [hunk 8]
```diff
@@ -55,6 +62,10 @@ public int shardId(Token t) {
     return (int) (sum >>> 32);
   }
 
+  public Integer getShardAwarePort() {
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/ShardingInfo.java:65]: Why not int?

File: driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java [hunk 9]
```diff
@@ -171,15 +183,33 @@ public void run() {
...
     this.host = host;
     this.hostDistance = hostDistance;
     this.manager = manager;
 
     this.minAllowedStreams = options().getMaxRequestsPerConnection(hostDistance) * 3 / 4;
 
-    this.timeoutsExecutor = manager.getCluster().manager.connectionFactory.eventLoopGroup.next();
+    this.scheduler = manager.getCluster().manager.connectionFactory.eventLoopGroup.next();
+
+    scheduler.schedule(
```
Comment (haaawk) [inline driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java:197]: Why do you need this rate-limiting? The driver opens at most connectionsPerShard connections to each shard at a time. Is it too many for you?

**Comment by haaawk** [comment]
Thanks again for your contribution @thirstycrow and sorry it took me so long to look at it.
It mostly looks ok. I took the liberty to split your work into smaller more manageable commits to make it easier for the reviewers.
I also addressed some of my concerns.
I uploaded the new version here -> https://github.com/scylladb/java-driver/pull/36
Please take a look.
